---
title: "Osteosarcoma"
output: github_document
---

```{r setup, include=FALSE}
library(rrrSingleCellUtils)
library(Seurat)
library(ggrepel)
library(tidyverse)
library(stringr)
library(harmony)
library(cowplot)
library(clustree)
set.seed(444)
```

## Download data from publications online
```{r}
#download from GSE152048
detail_subset <- read_tsv("sample_details.txt", show_col_types = FALSE) %>%
    filter(data_source == "GSE152048")

# Download, file, and extract the files from GEO
for (i in 1:nrow(detail_subset)) {                                     #nolint
    folder <- "input/downloads/"
    sample_name <- detail_subset$sample_name[i]
    site_link <- detail_subset$site_link[i]
    entire_path <- str_c(site_link, sample_name, ".matrix.tar.gz")
    tar_file <- str_c(folder, sample_name, ".tar.gz")
    system(paste0("wget ", entire_path, " -O ", tar_file))
    untar(tar_file, exdir = folder)
    file.remove(tar_file)
}


#download from GSE162454
detail_subset <- read_tsv("sample_details.txt", show_col_types = FALSE) %>%
    filter(data_source == "GSE162454")

for (i in 1:nrow(detail_subset)) {                                  #nolint
    folder <- "input/downloads/"
    tar_file <- str_c(folder, "GSE162454.tar.gz ")
    site_link <- detail_subset$site_link[i]
    sample_name <- detail_subset$sample_name[i]
    download.file(site_link, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = folder)
    file.remove(tar_file)
    bfile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_barcodes.tsv.gz") #nolint
    ffile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_features.tsv.gz") #nolint
    mfile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_matrix.mtx.gz") #nolint
    samp_dir <- str_c(folder, sample_name)
    dir.create(samp_dir)
    file.rename(str_c(folder, "/", bfile),
      str_c(samp_dir, "/", "barcodes.tsv.gz"))
    file.rename(str_c(folder, "/", ffile),
      str_c(samp_dir, "/", "features.tsv.gz"))
    file.rename(str_c(folder, "/", mfile),
      str_c(samp_dir, "/", "matrix.mtx.gz"))
  }



#download from GSE169396
detail_subset <- read_tsv("sample_details.txt", show_col_types = FALSE) %>%
    filter(data_source == "GSE169396")

for (i in 1:nrow(detail_subset)) {                                  #nolint
    folder <- "input/downloads/"
    tar_file <- str_c(folder, "GSE169396.tar.gz ")
    site_link <- detail_subset$site_link[i]
    sample_name <- detail_subset$sample_name[i]
    download.file(site_link, destfile = tar_file, method = "auto")
    untar(tar_file, exdir = folder)
    file.remove(tar_file)
    bfile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_barcodes.tsv.gz") #nolint
    ffile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_features.tsv.gz") #nolint
    mfile <- str_c(detail_subset$pre[detail_subset$sample_name == sample_name], sample_name, "_matrix.mtx.gz") #nolint
    samp_dir <- str_c(folder, sample_name)
    dir.create(samp_dir)
    file.rename(str_c(folder, "/", bfile),
      str_c(samp_dir, "/", "barcodes.tsv.gz"))
    file.rename(str_c(folder, "/", ffile),
      str_c(samp_dir, "/", "features.tsv.gz"))
    file.rename(str_c(folder, "/", mfile),
      str_c(samp_dir, "/", "matrix.mtx.gz"))
  }




```

## reading dataset from our reservoir and download
```{r}
set.seed(75454)
sample_details <- read_tsv("sample_details.txt", show_col_types = FALSE)

sample_details$sp_pattern <-
    str_replace(sample_details$sp_pattern, "nothing", "")

#Seurat objects and save individual qs object
obj <- parallel::mclapply(seq_len(nrow(sample_details)), function(i) {
    sample_name <- sample_details$sample_name[i]
    sample_path <- sample_details$sample_path[i]
    species <- sample_details$species[i]
    tumor_type <- sample_details$tumor_type[i]
    data_source <- sample_details$data_source[i]
    sp_pattern <- sample_details$sp_pattern[i]
    organism <- sample_details$organism[i]
    download_path <- "input/downloads/"
    ours_path <- "/gpfs0/home2/gdrobertslab/lab/Counts/"

    if (data_source == "ours") {
        path_10x <- paste0(path_10x = ours_path,
                           sample_path,
                           "/filtered_feature_bc_matrix/")
    } else {
        path_10x <- paste0(path_10x = download_path, sample_name, "/")
    }
        sobj <- tenx_load_qc(path_10x = path_10x,
                             species_pattern = sp_pattern,
                             violin_plot = FALSE) %>%
            process_seurat()
        sobj$sample_name <- sample_details$sample_name[i]
        sobj$tumor_type <- sample_details$tumor_type[i]
        sobj$data_type <- sample_details$data_type[i]
        sobj$model <- sample_details$model[i]
        sobj$organism <- sample_details$organism[i]
        sobj$location <- sample_details$location[i]
        qs::qsave(sobj, str_c("output/seurat_objects/single_sobj", "/",
                              species, "/",
                              tumor_type, "/",
                              sample_name,
                              organism, ".qs"))
    return(0)
}, mc.cores = 10)


# # individual QC numbers
# obj <- qs::qread("output/seurat_objects/single_sobj/xenograft/metastatic/S0254mouse.qs") #nolint

# # Cutoff and cleaning
# try_cutoff <- tribble(~feature,     ~min_val, ~max_val,
#                       "nCount_RNA", 0,        30000,
#                       "percent.mt", 0,        15)

# features <- c("nFeature_RNA", "nCount_RNA", "percent.mt")    #nolint

# # visualize the possible cutoff and adjust accordingly
# feature_hist(sobject = obj,
#              features = features,                         # nolint
#              cutoff_table = try_cutoff)

#quality control and downsampling and saving the merged sobj
what <- parallel::mclapply(unique(sample_details$unique), function(item) {
    tmp_df <- subset(sample_details, subset = (unique == item))
    temp_list <- list()
    for (i in seq_len(nrow(tmp_df))) {
        species <- tmp_df$species[i]
        tumor_type <- tmp_df$tumor_type[i]
        sample_name <- tmp_df$sample_name[i]
        group_obj <- tmp_df$unique[i]
        organism <- tmp_df$organism[i]
        sobj <- qs::qread(str_c("output/seurat_objects/single_sobj/",
                                species, "/",
                                tumor_type, "/",
                                sample_name,
                                organism, ".qs")) %>%
            subset(nCount_RNA < tmp_df$ncount_max[i] &
                percent.mt < tmp_df$mt_percent[i])
            temp_list[[sample_name]] <- sobj
            temp_list[[sample_name]] <-
                    subset(x = temp_list[[sample_name]],
                    cells = sample(Cells(temp_list[[sample_name]]),
                            min(4000, length(Cells(temp_list[[sample_name]])))))
    }
    merged_sobj <- merge(x = temp_list[[1]],
                                y = temp_list[2:length(temp_list)],
                                add.cell.id = names(temp_list)) %>%
                    process_seurat()
    qs::qsave(merged_sobj, str_c("output/seurat_objects/comb_sobjs/",
                                    item,
                                    ".qs"))
    run_harmony <- RunHarmony(object = merged_sobj,
                              group.by.vars = c("sample_name", "model"),
                              theta = c(3, 3))
    run_harmony <- process_seurat(sobject = run_harmony,
                                  reduction = "harmony")
    res_score <- optimize_silhouette(sobject = run_harmony,
                         test_res = seq(0.1, 0.9, by = 0.1),
                         summary_plot = FALSE) %>%
        filter(sil_vals == max(sil_vals)) %>%
        select(res_vals)
    run_harmony <- FindClusters(run_harmony,
                   resolution = res_score)
    qs::qsave(run_harmony, str_c("output/seurat_objects/harmony_sobjs/",
                                    item,
                                    ".qs"))
    return(item)
}, mc.cores = 10)

```



## Cell type annotaton function
```{r}
#loading the normal lungblood ref downloaded
human_lung <- qs::qread("input/downloads/normal_lung.qs")

human_lung$free_annotation <-
    str_replace_all(human_lung$free_annotation, c("/" = "_",
                                                  "\\+" = "_plus"))

human_lung$free_annotation <-
    stringr::str_replace_all(human_lung$free_annotation,
                              c("Alveolar Epithelial Type 1" = "Alveolar Epithelial",  #nolint
                                "Alveolar Epithelial Type 2" = "Alveolar Epithelial",  #nolint
                                "Basophil_Mast 1" = "Basophil_Mast",
                                "Basophil_Mast 2" = "Basophil_Mast",
                                "Bronchial Vessel 1" = "Bronchial Vessel",
                                "Bronchial Vessel 2" = "Bronchial Vessel",
                                "IGSF21_plus Dendritic" = "Dendritic cells",
                                "Myeloid Dendritic Type 1" = "Myeloid Dendritic",   #nolint
                                "Myeloid Dendritic Type 2" = "Myeloid Dendritic",   #nolint
                                "Natural Killer" = "NK cells",
                                "Natural Killer T" = "NKT",
                                "NK cells T" = "NKT",
                                "CD4_plus Memory/Effector T" = "CD4+ T cells",
                                "CD4_plus Naive T" = "CD4+ T cells",
                                "CD8_plus Memory_Effector T" = "CD8+ T cells",
                                "CD8_plus Naive T" = "CD8+ T cells",
                                "Proliferating NK_T" = "NKT",
                                "TREM2_plus Dendritic" = "Dendritic cells",
                                "Proliferating Macrophage" = "Macrophage",
                                "Nonclassical Monocyte" = "Monocytes",
                                "Classical Monocyte" = "Monocytes"))

#camilles paper mouse lung
mouse_lung <- qs::qread("input/downloads/mouse_lung.qs")

mouse_lung$CellType <-
            str_replace_all(mouse_lung$CellType, c("/" = "_",
                                                   "\\+" = "_plus",
                                                   "-" = "_"))

mouse_lung$CellType <-
    stringr::str_replace_all(mouse_lung$CellType,
                            c("Col14a1_plus fibroblast" = "Fibroblasts",
                              "Col13a1_plus fibroblast" = "Fibroblasts",
                              "AT1" = "AT",
                              "AT2 1" = "AT",
                              "AT2 2" = "AT",
                              "B cell 1" = "B cells",
                              "B cell 2" = "B cells",
                              "Cap_a" = "Cap",
                              "CD4 T cell 1" = "CD4",
                              "CD4 T cell 2" = "CD4",
                              "CD8 T cell 1" = "CD8",
                              "CD8 T cell 2" = "CD8",
                              "DC1" = "Dendritic cells",
                              "DC2" = "Dendritic cells",
                              "gd T cell" = "Tgd",
                              "Neut 1" = "Neutrophils",
                              "Neut 2" = "Neutrophils",
                              "NK cell" = "NK cells",
                              "Pericyte 1" = "Pericytes",
                              "Pericyte 2" = "Pericytes",
                              "Mast Ba2" = "Mast_Basophils",
                              "Mono" = "Monocytes"))

#function with mouse and human lung
annotate <- function(sobject,
                     species = "",
                     ref,
                     labels,
                     aggr_ref = FALSE,
                     label_type = "label.main",
                     ...) {
    if (species == "human") {
        hpca <- celldex::HumanPrimaryCellAtlasData()
        huim <- celldex::MonacoImmuneData()
        hpca$label.main <-
            stringr::str_replace_all(hpca$label.main,
                                     c("T_cells" = "T cells",
                                       "B_cell" = "B cells",
                                       "NK_cell" = "NK cells",
                                       "Monocyte" = "Monocytes",
                                       "DC" = "Dendritic cells"))
        ref <- list(hpca,
                    huim)
        labels <- list(hpca[[label_type]],
                      huim[[label_type]])
    } else if (species == "mouse") {
        mord <- celldex::MouseRNAseqData()
        moim <- celldex::ImmGenData()
        moim$label.main <-
            stringr::str_replace_all(moim$label.main,
                                    c("B cells, pro" = "B cells",
                                      "DC" = "Dendritic cells"))
        ref <- list(mord,
                    moim)
        labels <- list(mord[[label_type]],
                       moim[[label_type]])
    } else if (species == "human_lung") {
        hpca <- celldex::HumanPrimaryCellAtlasData()
        hpca$label.main <-
            stringr::str_replace_all(hpca$label.main,
                                     c("T_cells" = "T cells",
                                       "B_cell" = "B cells",
                                       "NK_cell" = "NK cells",
                                       "Monocyte" = "Monocytes",
                                       "DC" = "Dendritic cells"))
        huim <- celldex::MonacoImmuneData()
        ref3 <- GetAssayData(human_lung)
        ref <- list(hpca,
                    huim,
                    ref3)
        labels <- list(hpca[[label_type]],
                       huim[[label_type]],
                       human_lung$free_annotation)
    } else if (species == "mouse_lung") {
        mord <- celldex::MouseRNAseqData()
        moim <- celldex::ImmGenData()
        ref3 <- GetAssayData(mouse_lung)
        moim$label.main <-
            stringr::str_replace_all(moim$label.main,
                                    c("B cells, pro" = "B cells",
                                      "DC" = "Dendritic cells"))
        ref <- list(mord,
                    moim,
                    ref3)
        labels <- list(mord[[label_type]],
                       moim[[label_type]],
                       mouse_lung$CellType)
    }
    annotation <-
        SingleR::SingleR(test = Seurat::as.SingleCellExperiment(sobject),
                         ref = ref,
                         labels = labels,
                         aggr.ref = aggr_ref,
                         ...)
    sobject$annotations <- annotation$labels
    sobject$cell_scores <-
        apply(X = annotation$scores,
              MARGIN = 1,
              function(x) max(x, na.rm = TRUE))
    return(sobject)
}

```


## Annotate all the datasets
```{r}
#make table for what ref to use
refs <- tribble(~sobj,            ~ref,
               "patient_prim",    "human",
               "normal_bone",     "human",
               "patient_mets",    "human_lung",
               "xeno_prim_human", "human",
               "xeno_mets_human", "human",
               "xeno_prim_mouse", "mouse",
               "xeno_mets_mouse", "mouse",
               "mm_prim",         "mouse",
               "mm_mets",         "mouse_lung")

#annotate using the function parallelly
what <- parallel::mclapply(seq_len(nrow(refs)), function(i) {
    sobj <- refs$sobj[i]
    ref <- refs$ref[i]
    object <- qs::qread(str_c("output/seurat_objects/harmony_sobjs/",
                              sobj,
                              ".qs"))
    ann_sobj <- annotate(sobject = object,
                         species = ref,
                         aggr_ref = TRUE)
    cluster_celltypes <-
        table(ann_sobj$seurat_clusters, ann_sobj$annotations) %>% 
        as.data.frame() %>%
        group_by(Var1) %>%
        arrange(desc(Freq), .by_group = TRUE) %>%
        slice_head(n=1)
    ann_sobj$new <- ann_sobj$seurat_clusters
    for (item in seq_len(nrow(cluster_celltypes))) {
        seurat_clust <- str_c("^", cluster_celltypes$Var1[item], "$") %>%
            as.character()
        celltype <- cluster_celltypes$Var2[item] %>%
            as.character()
        ann_sobj$new <- str_replace_all(string = ann_sobj$new,
                                        pattern = seurat_clust,
                                        replacement = celltype)
    }
    # wanted_cells <- table(ann_sobj$annotations) %>%
    #     as.data.frame() %>%
    #     filter(Freq > 30)
    # Idents(ann_sobj) <- ann_sobj$annotations
    # filtered_ann_sobj <- subset(x = ann_sobj,
    #                             idents = wanted_cells$Var1)
    qs::qsave(ann_sobj, str_c("output/seurat_objects/harmony_sobjs/",
                                       sobj,
                                       ".qs"))
    print(str_c(ref, " ", "annotated", " ", sobj))
}, mc.cores = 10)

```


## SNVs calling
```{r}
#Using snvs to differentiate cancer cells
#for (item in unique(sample_details$unique)) {
temp_df <- subset(sample_details, subset = (unique == "patient_mets"))
sobj <- qs::qread("output/seurat_objects/harmony_sobjs/patient_mets.qs")
control_celltypes <- c("Dendritic cells", "Macrophage")
snv_sobjs <- list()
for (i in seq_len(nrow(temp_df))) {
    sample_path <- temp_df$sample_path[i]
    sample_name <- temp_df$sample_name[i]
    data_type <- temp_df$data_type[i]
    Idents(sobj) <- sobj$sample_name
    new_sobj <- subset(sobj, ident = sample_name)
    if (grepl("^S0", sample_name)) {
        normal_clusters <-
            match_celltype_clusters(sobject = new_sobj,
                                    normal_celltypes = control_celltypes,
                                    cluster_col = "seurat_clusters",
                                    celltype_col = "new")
        c_b_t <-
            new_sobj@meta.data %>%
            select(seurat_clusters) %>%
            dplyr::rename(cell_group = seurat_clusters) %>%
            rownames_to_column("cell_barcode") %>%
            as_tibble()

        bamfile_name <- if_else(data_type == "three_prime",
                                "possorted_genome_bam.bam",
                                "gex_possorted_bam.bam")

        c_b_t <- c_b_t %>%
            mutate(bam_file = paste0("/home/gdrobertslab/lab/Counts/",
                                     sample_path,
                                     "/",
                                     bamfile_name),
                   cell_barcode = str_remove(cell_barcode, ".+_"))

        snp_tree <-
            get_snp_tree(cellid_bam_table = c_b_t,
                            ploidy = "GRCh37",
                            ref_fasta = "/home/gdrobertslab/lab/GenRef/10x-human/fasta/genome.fa",     #nolint
                            min_depth = 5,
                            temp_dir = paste0("output/snv_calling/temp_dir/",
                                              sample_name),
                            sbatch_base = "new_sobj",
                            slurm_base = paste0("output/snv_calling/slurmOut/",   #nolint
                                                sample_name),
                            min_sites_covered = 1000,
                            cleanup = FALSE)
        png(str_c("output/snv_calling/snv_trees/",
                    sample_name, ".png"),
            width = 2500,
            height = 2500,
            res = 300)
        plot(snp_tree)
        dev.off()

        snp_sobj <-
            label_tree_groups(sobject = new_sobj,
                                dist_tree = snp_tree,
                                group_col_name = "seurat_clusters",
                                normal_groups = normal_clusters,
                                cut_n_groups = 2)
        
        snv_sobjs[[sample_name]] <- snp_sobj

        umap_plot <- DimPlot(object = snp_sobj,
                group.by = c("snp_tumor_call",
                             "new"),
                label = TRUE,
                repel = TRUE) +
            NoLegend()
        ggsave(str_c("output/snv_calling/", sample_name, "_umap.pdf"),
        width = 4,
        height = 3,
        plot = umap_plot)

    } else {
        print(str_c(sample_name, " does not have a bam file"))
    }
}

 dist_tree <-
        calc_tree(matrix_file = "output/snv_calling/temp_dir/S0058/distances.tsv",
                  counts_file = "output/snv_calling/temp_dir/S0058/distances_tot_count.tsv",
                  min_sites = 5)

file.exists("allsamples.rmd")                                                                                    
[1] FALSE
> file.exists("Rfiles/allsamples.rmd")

````


## Split cancer vs the stroma - here
label potential tumor cell and then split the object, run harmony and save

```{r}
refs <- tribble(~sobj,            ~split,   ~has_multiomics,
               "patient_prim",    "yes",    "no",
               "normal_bone",     "no",     "no",
               "patient_mets",    "yes",    "yes",
               "xeno_prim_human", "no",     "yes",
               "xeno_mets_human", "no",     "no",
               "xeno_prim_mouse", "no",     "yes",
               "xeno_mets_mouse", "no",     "no",
               "mm_prim",         "yes",    "no",
               "mm_mets",         "yes",    "no")

cancer_celltypes <- list(patient_prim = c("iPS_cells",
                                         "MSC",
                                         "Neurons",
                                         "Smooth_muscle_cells",
                                         "Tissue_stem_cells",
                                         "Fibroblasts",
                                         "Chondrocytes",
                                         "Osteoblasts"),
                        normal_bone = c(),
                        patient_mets = c("Myofibroblast",
                                         "MSC",
                                         "Neurons",
                                         "Tissue_stem_cells",
                                         "Osteoblasts",
                                         "Fibromyocyte",
                                         "Chondrocytes",
                                         "iPS_cells",
                                         "Pericyte",
                                         "Smooth_muscle_cells",
                                         "Airway Smooth Muscle",
                                         "Alveolar Fibroblast",
                                         "Adventitial Fibroblast"),
                        xeno_prim_human = c(),
                        xeno_mets_human = c(),
                        xeno_prim_mouse = c(),
                        xeno_mets_mouse = c(),
                        mm_prim = c("Fibroblasts"),
                        mm_mets = c("Fibroblasts"))

tumor_vs_stroma <- parallel::mclapply(seq_len(nrow(refs)), function(item) {
    sobj <- refs$sobj[item]
    split <- refs$split[item]
    multiomics <- refs$has_multiomics[item]
    object <- qs::qread(str_c("output/seurat_objects/harmony_sobjs/",
                              sobj,
                              ".qs"))
    for (cell in cancer_celltypes[[sobj]]) {
        object$annotations <- str_replace_all(string = object$annotations,
                                              pattern = cell,
                                              replacement = "cancer_cells")}
    qs::qsave(object, str_c("output/seurat_objects/annotated_sobjs/",
                            sobj,
                            ".qs"))
    if (split == "yes") {
        normal_cells <- table(object$annotations) %>%
        as.data.frame() %>%
        filter(Var1 != "cancer_cells")
        Idents(object) <- object$annotations
        normal_obj <- subset(x = object,
                             idents = normal_cells$Var1) %>%
            process_seurat()
        normal_obj <- RunHarmony(object = normal_obj,
                                 group.by.vars = "sample_name")
        normal_obj <- normal_obj %>%
            process_seurat(reduction = "harmony")
        qs::qsave(normal_obj, str_c("output/seurat_objects/tumor_vs_stroma/",   #nolint
                                    sobj,
                                    "_normal",
                                    ".qs"))
        Idents(object) <- object$annotations
        cancer_obj <- subset(x = object,
                             idents = "cancer_cells")
        if (multiomics == "yes") {
            cancer_obj <- RunHarmony(object = cancer_obj,
                                     group.by.vars = c("sample_name", "model", "data_type"),  #nolint
                                     theta = c(3, 3, 3))
        } else {
            cancer_obj <- RunHarmony(object = cancer_obj,
                                     group.by.vars = c("sample_name", "model"),
                                     theta = c(3, 3))
        }
        cancer_obj <- cancer_obj %>%
            process_seurat(reduction = "harmony",
                           resolution = 0.05)
        qs::qsave(cancer_obj, str_c("output/seurat_objects/tumor_vs_stroma/",  #nolint
                                    sobj,
                                    "_tumor",
                                    ".qs"))
    } else {
        qs::qsave(object, str_c("output/seurat_objects/tumor_vs_stroma/",
                                sobj,
                                ".qs"))
    }
    print(str_c(sobj, " labelled with cancer cells and subsetted"))
}, mc.cores = parallelly::availableCores())


```

## DEGs and GSEA plot

```{r}
mm_prim_tumor <- qs::qread("output/seurat_objects/harmony_sobjs/patient_mets.qs")

DimPlot(mm_prim_tumor, group.by = c("sample_name", "seurat_clusters", "new"))

run_degs <- function(sobject,
                     group_by = "",
                     prefix = "") {
    if (group_by == "seurat_clusters") {
        sobject$pseudobulk_col <- stringr::str_c(sobject$sample_name,
                                 "-",
                                 sobject$seurat_clusters)
    } else if (group_by == "annotations") {
        sobject$annotations <-
            stringr::str_replace_all(sobject$annotations, " ", "")
        sobject$pseudobulk_col <- stringr::str_c(sobject$sample_name,
                                 "-",
                                 sobject$annotations)
    }
    pseudobulk <- Seurat::AggregateExpression(object = sobject,
                                      group.by = "pseudobulk_col",
                                      slot = "counts")[[1]]
    deseq_coldata <- data.frame(pseudobulk_group = colnames(pseudobulk))
    rownames(deseq_coldata) <- deseq_coldata$pseudobulk_group
    deseq_coldata <- tidyr::separate(data = deseq_coldata,
                          col = "pseudobulk_group",
                          into = c("sample_name", "group"),
                          sep = "-")
    deseq_coldata$type <- "single-read"
    deseq_coldata <- deseq_coldata[, -1]
    if (group_by == "seurat_clusters") {
        deseq_group <- levels(factor(sobject$seurat_clusters))
        names(deseq_group) <- paste0(prefix, "_Cluster_", deseq_group) #nolint
    } else if (group_by == "annotations") {
        deseq_group <- levels(factor(sobject$annotations))
        names(deseq_group) <- paste0(prefix, "_Celltype_", deseq_group) #nolint
    }
    degs_output <- list()
    for (group in names(deseq_group)) {
        interested_group <- deseq_group[group]
        other_group <- deseq_coldata
        other_group$group[other_group$group == interested_group] <- "target"  #nolint
        other_group$group[other_group$group != "target"] <- "other"         #nolint
        degs_output[[group]] <-
            DESeq2::DESeqDataSetFromMatrix(countData = pseudobulk,
                                           colData = other_group,
                                           design = ~ group) %>%
            DESeq2::DESeq() %>%
            DESeq2::results() %>%
            as.data.frame() %>%
            dplyr::arrange(desc(log2FoldChange)) %>%
            tibble::rownames_to_column("gene") %>%
            na.omit()
    }
    return(degs_output)
}


run_gsea <- function(degs_result,
                     category = "",
                     subcategory = "",
                     species = "") {
    gsea_ref <- msigdbr::msigdbr(species = species,
                                 category = category,
                                 subcategory = subcategory) %>%
        split(x = .$gene_symbol, f = .$gs_name)

    gsea_output <- list()
    for (item in names(degs_result)) {
        gsea_input <- as.vector(degs_result[[item]]$log2FoldChange)
        names(gsea_input) <- degs_result[[item]]$gene

        output <- fgsea::fgseaMultilevel(gsea_ref,
                                         gsea_input,
                                         minSize = 10,
                                         maxSize = 500,
                                         nPerm = 1000)
        gsea_output[[item]] <- output %>%
            dplyr::arrange(desc(NES)) %>%
            filter(padj < 0.05) %>%
            na.omit()
    }
    return(gsea_output)
}

#maybe write a tsv of the output

````

## Plots

```{r}

# theme_roberts <- function(axis_font_size = 5,
#                           axis_title_font_size = 8,
#                           facet_font_size = 5,
#                           font_name = "Arial",
#                           legend_key_size = "0.2",
#                           legend_text_font_size = 5,
#                           legend_title_font_size = 6,
#                           subtitle_font_size = 6,
#                           title_font_size = 10) {
#     list(ggpubr::theme_pubr(base_family = font_name) +
#          ggplot2::theme(
#             axis.text = ggplot2::element_text(size = axis_font_size),
#             axis.title = ggplot2::element_text(size = axis_title_font_size),
#             plot.title = ggplot2::element_text(size = title_font_size,
#                                                face = "bold",
#                                                hjust = 0.5),
#             strip.text = ggplot2::element_text(size = facet_font_size,
#                                                face = "bold"),
#             plot.subtitle = ggplot2::element_text(hjust = 0.5,
#                                                   size = subtitle_font_size),
#             legend.text = ggplot2::element_text(size = legend_text_font_size),
#             legend.title = ggplot2::element_text(size = legend_title_font_size,
#                                                  face = "bold"),
#             strip.background = ggplot2::element_blank(),
#             legend.key.size = ggplot2::unit(0.2, "cm")
#             ),
#     ggplot2::scale_color_manual(values = plot_cols))
# }

two_way_plot <- function(data, x_col = "z_score") {
    lab4plot <-
        tibble(y = c(-3, 3),
               x = c(0.2, 0.2),
               label = c("Downregulated", "Upregulated"))

    plot_name <-
        ggplot() +
        geom_bar(data = data,
                 aes(x = -1 * order,
                     y = get(x_col),
                     fill = get(x_col) > 0),
                 stat = "identity",
                 alpha = 0.8) +
        coord_flip() +
        geom_hline(yintercept = 0,
                    color = "black",
                    linewidth = 0.5) +
        geom_text(data = data,
                    aes(x = -1 * order,
                        y = y_pos,
                        hjust = justify_y,
                        label = pathway),
                    size = 1) +
        geom_text(data = lab4plot,
                  aes(x = x,
                      y = y,
                      label = label),
                  fontface = "bold",
                  size = 2) +
        scale_fill_manual(values = plot_cols,
                          name = paste0(x_col, " > 0")) +
        theme(strip.background = element_rect(color = "white",
                                              fill = "white"),
              legend.position = "none",
              panel.grid.major = element_blank(),
              panel.grid.minor = element_blank(),
              panel.border = element_blank(),
              axis.line.x = element_line(color = "black"),
              axis.line.y = element_blank(),
              axis.title.x = element_text(size = 7),
              axis.ticks.y = element_blank(),
              plot.title = element_text(size = 10, face = "bold"),
              axis.text.y = element_blank(),
              axis.text.x = element_text(size = 7)) +
        labs(title = data$sample[1],
             y = "NES",
             x = "") +
        theme(plot.title = element_text(hjust = 0.5)) +
        ylim(-6, 6)

    return(plot_name)
}


degs_result <- run_degs(sobject = mm_prim_tumor,
                        group_by = "seurat_clusters",
                        prefix = "mm_prim_tumor")

gsea_result <- run_gsea(degs_result = degs_result,
                        category = "C2",
                        subcategory = "CP:KEGG",
                        species = "mouse")


for (item in names(gsea_result)) {
    top5up_down <- rbind(gsea_result[[item]] %>%
                            filter(NES > 0) %>%
                            slice_head(n = 5),
                         gsea_result[[item]] %>%
                            filter(NES < 0) %>%
                            slice_tail(n = 5)) %>%
                    dplyr::select(pathway,
                                    NES,
                                    size,
                                    padj) %>%
                    arrange(desc(NES)) %>%
                    mutate(pathway = as.factor(pathway) %>%
                            str_wrap(80) %>%
                            fct_reorder(NES),
                       sample = item,
                       order = 1:n(),
                       justify_y = if_else(NES > 0, 1, 0),
                       y_pos = if_else(NES > 0, -0.1, 0.1))
    pdf(paste0("output/plots/mousemodel/gsea/gsea_",
                   item,
                   "_barplots.pdf"),
            width = 4,
            height = 3)
        print(two_way_plot(top5up_down, x_col = "NES"))
        dev.off()
}

```





